Issue 706 - New attribute for test fixtures/cases supporting external data sources

acco32 opened the pull request:
***Do not merge***

Just a checkpoint review. Some items I still need to deal with are removing the test files once the suite (fixture) is run. The only remaining piece is to add the database connection string implementation. I thought this would be a good place to stop and get some feedback. Thanks in advance.

Fixes #706

CharliePoole commented:
The .NET 4.5 tests don't seem to be able to find the files.

acco32 commented:
Perhaps they are not being copied fast enough from the binary. I was having  some trouble deleting the files after the test. This was something I was hoping to get some help.

Outside of the files, any other comments on the implementation? Are you able to run the tests locally?

CharliePoole commented:
I'm tied up in another city, so I won't be able to do a true review till
the weekend.

On Thu, Sep 24, 2015 at 8:05 PM, Matthew Moore <notifications@github.com>
wrote:

> Perhaps they are not being copied fast enough from the binary. I was
> having some trouble deleting the files after the test. This was something I
> was hoping to get some help.
>
> Outside of the files, any other comments on the implementation? Are you
> able to run the tests locally?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nunit/nunit/pull/843#issuecomment-143109936>.
>

rprouse commented on nunit.sln:
```diff
@@ -1,6 +1,9 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
 # Visual Studio 2012
+# SharpDevelop 5.1
```
Did you make any actual changes to the solution or is this just SharpDevelop re-arranging things? Any added projects?

acco32 replied:
This is just Sharpdevelop I believe. I like it as it is lightweight compared to [VS Code](https://en.wikipedia.org/wiki/Visual_Studio_Code) is looking interesting.

rprouse commented:
I updated the title and added to your PR description to link to the issue.

rprouse commented on src/NUnitFramework/framework/Attributes/DataSourceAttribute.cs:
```diff
@@ -0,0 +1,392 @@
+  /// <summary>
+  /// The type of external data source
+  /// </summary>
+  public enum DataSourceType: int {
```
Minor, but we normally don't specify the enum type (int) if it is the default and the bracket should be on a new line.

rprouse commented on src/NUnitFramework/framework/Attributes/DataSourceAttribute.cs:
```diff
@@ -0,0 +1,392 @@
+  /// </summary>
+  public enum DataSourceType: int {
+    /// <summary>
+    /// No datas ource is specified. This is the default state.
```
Comment needs fixing "No datas ource"

rprouse commented on src/NUnitFramework/framework/Attributes/DataSourceAttribute.cs:
```diff
@@ -0,0 +1,392 @@
+    /// <summary>
+    /// No datas ource is specified. This is the default state.
+    /// </summary>
+    Null = 0,
```
Glad you have a default value, but I would prefer something like 'None' here. Null is a bit of an overloaded term.

rprouse commented on src/NUnitFramework/framework/Attributes/DataSourceAttribute.cs:
```diff
@@ -0,0 +1,392 @@
+    /// <param name="dataType">The type of data provider</param>
+    /// <param name="sourceName">The connection string for the data source. It could be a filename or DB connection string.</param>
+    /// <param name="rowCount">The number of rows to collect from the data source. By default it is all rows</param>
+    public DataSourceAttribute(DataSourceType dataType, string sourceName, int rowCount = (int)DataRowCount.All)
```
I am curious how others feel about this, but I don't like using an enum for the rowCount default value and casting to an int, especially since the enum only has one value. I think I would prefer a `public const int AllRows = 0` on the class.

CharliePoole replied:
I would make RowCount a named paramter and remove it from the constructor. Set the backing store to -1 in the constructor so you can tell if it was set. If it's not set, get all rows. Generally, it's best to use a named rather than a positional parameter for optional args to an Attribute class.

rprouse commented on src/NUnitFramework/framework/Attributes/DataSourceAttribute.cs:
```diff
@@ -0,0 +1,392 @@
+              {
+                Array array = item as Array;
+
+                #if NETCF
```
Another minor formatting issue, we put the #if statements in column 0 so they are obvious.

rprouse commented on src/NUnitFramework/framework/Attributes/DataSourceAttribute.cs:
```diff
@@ -0,0 +1,392 @@
+      
+      try {
+        
+        if (DataType == DataSourceType.Csv){
```
I would prefer a switch statement here. As DataSourceType grows in the future, it would be easier to add to and we can log errors in the default or throw or something. At the very lease, it should be an if/else if block.

acco32 replied:
I tend to avoid switch statements in favour of if/else blocks as I find if/else is more flexible

rprouse replied:
You should really have some sort of data source factory class that constructs a data source based on the enum type and returns a base DataSource that you call GetData on. These higher level classes should not be so tightly bound to the underlying data sources. Adding a new data source should be as simple as deriving a new DataSource class, implementing its GetData method and then adding it to the enum and factory. The factory can also be tested to ensure that it handles every enum value. That is difficult to do in this class because it is so tightly coupled with so many things.

rprouse commented on src/NUnitFramework/framework/Attributes/DataSourceAttribute.cs:
```diff
@@ -0,0 +1,392 @@
+        }
+        
+      } catch (Exception e) {
+        throw e;
```
If you don't end up removing this, you should never use `throw e` in situations like this. Instead, just use `throw`. The way you did it, the callstack gets rewritten to the current location, if you just throw, the callstack gets preserved.

If this stays in, we should add some logging here.

acco32 replied:
Didn't know that. Will adjust. Thanks.

rprouse commented on src/NUnitFramework/framework/Attributes/DataSourceAttribute.cs:
```diff
@@ -0,0 +1,392 @@
+      #region Error Checking
+      
+      // file name sources
+      if (DataType == DataSourceType.Csv || DataType == DataSourceType.Xml){
```
Please follow our coding standards, all braces on newlines and indent by 4 spaces.

rprouse replied:
If we are considering supporting more file based datatypes in the future, you should probably add an IsFileBasedDataSource read-only property to be used throughout this class. For example, I could see a straight text file with one value per line (or is this just a CSV with one column?).

acco32 replied:
I'd probably consider your use case to be a CSV with a single column. If I separate the source types to individual classes, could we not differentiate by type instead of the property?

rprouse commented on src/NUnitFramework/framework/Attributes/DataSourceAttribute.cs:
```diff
@@ -0,0 +1,392 @@
+      
+      // file name sources
+      if (DataType == DataSourceType.Csv || DataType == DataSourceType.Xml){
+        if (!File.Exists(SourceName)){
```
This if can be combined with the previous one.

rprouse commented on src/NUnitFramework/framework/Attributes/DataSourceAttribute.cs:
```diff
@@ -0,0 +1,392 @@
+      return null;
+    }
+    
+    private IEnumerable GetCsvData(IMethodInfo method)
```
I would prefer to see the actual parsing of CSV or XML data out to helper classes that can be tested directly rather than through this attribute.

CharliePoole replied:
+1

rprouse replied:
You have an Abstract DataSource base class. That class should have an abstract `GetData` method then each of these methods can be moved to the `DataSource` implementations.

rprouse commented on src/NUnitFramework/framework/nunit.framework-4.5.csproj:
```diff
@@ -123,6 +123,7 @@
     <Compile Include="Attributes\DatapointAttribute.cs" />
     <Compile Include="Attributes\DatapointsAttribute.cs" />
     <Compile Include="Attributes\DatapointSourceAttribute.cs" />
+    <Compile Include="Attributes\DataSourceAttribute.cs" />
```
Obviously, this will need to be added to all the projects once it is complete.

CharliePoole replied:
Even before that, I'd say. :-) We'd like to see the tests passing for all projects.

CharliePoole replied:
Oh! I suppose this is why we are only failing in the .NET 4.5 build - the code is not in the other builds yet!

rprouse commented:
I am about half way through my code review. I will get back to it later when I have a bit more time.

CharliePoole commented on src/NUnitFramework/framework/Attributes/DataSourceAttribute.cs:
```diff
@@ -0,0 +1,392 @@
+    /// <summary>
+    /// The data input type
+    /// </summary>
+    public DataSourceType DataType {get; private set;}
```
I dislike XxxxType as a name pattern when it's not a System.Type.

An alternative to having this enum as an argument would be to derive three different attributes for Csv, Xml and database respectively.

CharliePoole commented on src/NUnitFramework/framework/Attributes/DataSourceAttribute.cs:
```diff
@@ -0,0 +1,392 @@
+
+        if(source != null)
+        {
+          #if NETCF
```
Have you tested on CF?

acco32 replied:
I've only tested on the desktop. I took an existing attribute to learn the structure and then modified it accordingly. I'm trying to use as many generic calls as I can for it to be compatible on different framework versions.

CharliePoole commented on src/NUnitFramework/framework/Attributes/DataSourceAttribute.cs:
```diff
@@ -0,0 +1,392 @@
+          }
+        }
+      }
+      catch (Exception ex)
```
The above code seems to duplicate what we have elsewhere to a great extent. Can we refactor to avoid the duplication?

CharliePoole commented on src/NUnitFramework/framework/Attributes/DataSourceAttribute.cs:
```diff
@@ -0,0 +1,392 @@
+      var dataRows = xmlDoc.SelectSingleNode("//data").ChildNodes;
+      int N = DataRows != 0 ? DataRows : dataRows.Count;
+      
+      for (int i = 1; EndOfData(i,N);i++){
```
Use of EndOfData in the loop seems to make the code somewhat obscure, especially since EndOfData returns true when we are NOT at the end of the data. Spelling out the test would make the for statement slightly more complex but probably easier to understand.

CharliePoole commented on src/NUnitFramework/tests/TestXmlFile1.xml:
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="utf-8"?>
```
All the test XML files use the same element names. Are those names required?

acco32 replied:
Right now, the top level root node is hard coded to *data*. The child nodes can have various names, so long as the column count matches that of the method's argument count

CharliePoole commented:
General comments:

1. I'm not 100% satisfied with the name of the attribute. Using Source makes sense, since it's a source of data just like a TestCaseSourceAttribute. But "DataSource" seems a bit general. A fully meaningful name could be ExternalTestCaseSourceAttribute, but that's a real mouthful! Let's think about this.

2. One possibility would be to create three different attributes, possibly derived from a common one or possibly just using some common helper classes. For example, XmlTestCaseSourceAttribute, etc.

3. Another option would be **not** to have new attributes but to use TestCaseSource with some built-in sources we provide. For example [TestCaseSource(typeof(XmlDataSource), "filename")]. Then all the TestCaseSource code would automatically be shared.

4. A question: what databases will the Db option work with as written?

All in all, this is looking pretty good. Let's figure out what directions we want to take it. Thanks for your work so far and to come, @acco32 !

Charlie

acco32 commented:
Perhaps XmlData, DatabaseData, CsvData? Would external be implied based on the remaining attribute usage?

Perhaps combining everything into TestCaseSource can be the next iteration once we have the individual attributes ready.

The databases supported are those supported by the internal ODBC interface. The connection strings are readily available for numerous vendors (MSSQL, MySql, Oracle, MS Access). Others will require third party library imports which is probably handled better as an import module of that respective library.

@rprouse, @CharliePoole, I appreciate the feedback. Let me know if you have other suggestions. Also, what are your thoughts on JSON source? This too can be done internally (System.Web I believe), but other libraries exists that I'd rather use. This comes back to the question of supporting third party systems. 

MM

acco32 commented:
@CharliePoole , @rprouse 

Now getting time to get back to this and apply the recommendations and do further development. A problem I'm having is getting all the test files to get written to disk before the tests execute. I have a one time setup function to do this task, but it seems the test workers are trying to access the data before it is created. On subsequent attempts it works every time as the data is flushed.

I am looking for some options to ensure the data is written to disk before the tests execute. I've looked at adding sleep statements and even specifying the number of workers to one to no avail. Both of which I want to avoid.

Any recommendations?

CharliePoole commented:
Hi Matthew,

Rob and I are in the middle of getting out NUnit 3.0.1 as well as CTP-8 of the VS adapter. I'll take a look at this again in a day or two.

Charlie

CharliePoole commented:
@acco32 I suggest using the TestFile class for this purpose.
  https://github.com/nunit/nunit/blob/master/src/NUnitFramework/tests/TestUtilities/TestFile.cs

If you search in the test project, you'll find it used in a few places.

rprouse commented:
@acco32 sorry, we have let this drop. Sometimes you need to give us prods to let us know you have checked in more changes and would like another review.

Can you update me with the current status of your code, then I will review if you are still interested in getting this PR done and merged.

acco32 commented:
I just finished the implementation and will submit a pull request very soon. Sone more testing on my local server needs to be done but I can enter and parse connection stings and use the appropriate driver to submit a query. Will definitely want a review. Thanks for the follow up.

 
 
  On Thu, Feb 4, 2016 at 8:40, Rob Prouse<notifications@github.com> wrote:   
@acco32 sorry, we have let this drop. Sometimes you need to give us prods to let us know you have checked in more changes and would like another review.

Can you update me with the current status of your code, then I will review if you are still interested in getting this PR done and merged.

—
Reply to this email directly or view it on GitHub.

acco32 commented:
**Do not merge**

@CharliePoole, @rprouse 

Another checkpoint review (based on previous feedback). Finished DB connection object with a test file. Some further improvements from this point would be better code organization.

This was a little harder to test that originally thought. The connections strings can be referenced from [http://www.dofactory.com/reference/connection-strings]. Not all are included and some more thorough testing will be required third party DBs (SQLExpress, MySql, etc...). Right now they are manual. I was originally thinking to tap into a public DB, but opted not.

acco32 commented:
Also, for the travis build, it seems the ODBC driver is missing causing the build to fail.

acco32 commented:
@rprouse, @CharliePoole 

do you guys have time to review?

rprouse commented:
@acco32 I will continue the review as soon as possible. I added a reminder to Todoist to prod me :smile: 

Meanwhile, have you read through the comments for the review I have already done? Also, for testing, it would be better to abstract out the need for ODBC. I think database access should be injected and mockable.

acco32 commented:
Thanks. I did read the original comments and implemented what I could in the time I had. Been a bit busy recently as I just changed positions. As I now have the objects in separate object, mocking is possible and should be straightforward. With that said, I still want a critique on the current organization.

 
 


 On Wed, Feb 24, 2016 at 6:52, Rob Prouse<notifications@github.com> wrote:  @acco32 I will continue the review as soon as possible. I added a reminder to Todoist to prod me  Meanwhile, have you read through the comments for the review I have already done? Also, for testing, it would be better to abstract out the need for ODBC. I think database access should be injected and mockable.—
Reply to this email directly or view it on GitHub.

acco32 commented:
@CharliePoole, @rprouse 

I've been looking at the code and need some suggestions in testing the attributes. The code works on any windows based OS. Have a look and let me know how best to refactor to pass the travis build. Happy Easter.

rprouse commented on src/NUnitFramework/framework/Attributes/CsvDataSourceAttribute.cs:
```diff
@@ -0,0 +1,206 @@
+                      parameters = mi == null ? new ParameterInfo[0] : mi.GetParameters();
+                    }
+#endif
+                                if (args.Length != parameters.Length)
```
Your indentation is all over the place. Do you should have replace tabs with places turned on and tabs/indents should be 4. It makes it very hard to read and review.

rprouse commented on src/NUnitFramework/framework/Attributes/CsvDataSourceAttribute.cs:
```diff
@@ -0,0 +1,206 @@
+    }
+
+
+
```
I may be a bit OCD, but multiple empty lines when they are not needed are a pet peeve of mine :smile:

rprouse commented on src/NUnitFramework/framework/CsvData.cs:
```diff
@@ -0,0 +1,136 @@
+
+            if (rowNum <= 0)
+            {
+                throw new Exception("Number of rows to read cannot be zero or less");
```
You should be throwing an ArgumentException and specifying the argument in question. Also, since you are throwing, you do not need the else below.

rprouse commented on src/NUnitFramework/framework/CsvData.cs:
```diff
@@ -0,0 +1,136 @@
+        /// </summary>
+        /// <param name="source">CSV file</param>
+        /// <param name="rowNum">number of rows to read</param>
+        public CsvData(string source, int rowNum)
```
I think the `rowNum` argument and the `RowNumber` property are confusing. If I did not read the comment, I would have assumed it was which row you want to read, not the number of rows. I think it should be `numRows` and `NumberOfRows`. When designing a public API, it is important to use clear, concise names. @CharliePoole and I often go back and forth over naming endlessly :smile:

rprouse commented on src/NUnitFramework/framework/Attributes/DataSourceAttribute.cs:
```diff
@@ -0,0 +1,392 @@
+    /// <summary>
+    /// Data source is a database
+    /// </summary>
+    Db = 4
```
Rename this to `Database`

rprouse commented on src/NUnitFramework/framework/Attributes/DatabaseDataSourceAttribute.cs:
```diff
@@ -0,0 +1,161 @@
+    /// Specifies an external data source for test methods. This is evaulated at compile time.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Method, AllowMultiple = true, Inherited = false)]
+    class DbDataAttribute : NUnitAttribute, ITestBuilder, IImplyFixture
```
Class name should match the filename and I prefer the full `DatabaseDataSourceAttribute`.

rprouse commented:
When you are testing your attributes, you should not only be using them, but you need tests that construct the attributes like you would any other class and exercise/test their methods directly. You should also be excluding/including tests by platform using the Platform attribute. That might require that you break up some of your tests so that you can apply the Platform attribute where needed.

I would also ask that you go through your changes, clean up the formatting and start doing some of the refactoring that @CharliePoole and I have asked for. There are getting to be so many comments that it is getting hard to review. I know we are asking for many changes and we have high standards, but this is a very large open source project with many contributors. It is important that we keep the code as clean and maintainable as possible.

We are happy to keep working with you on this and you are moving in the right direction, but this is a big enhancement and we want to get it right, so there will be a lot of back and forth on it.

CharliePoole commented:
@acco32 You have a lot of code here that duplicates TestCaseSourceAttribute. What is more, the latter has had various fixes over time, which may need to be replicated here... forever!

Could the three non-abstract attributes simply derive from TestCaseSourceAttribute? Or from an abstract base that we split out of TestCaseSourceAttribute?

I realize that this could seem like starting over, but it isn't really because the programmer is always made smarter by the code he writes, whether it gets used as is or not. IOW I think it could go pretty fast if we did a reset.

acco32 commented:
Thanks for the review. I'll improve the code and request another review to merge when ready.

 
 
  On Sun, Mar 27, 2016 at 19:19, CharliePoole<notifications@github.com> wrote:   
@acco32 You have a lot of code here that duplicates TestCaseSourceAttribute. What is more, the latter has had various fixes over time, which may need to be replicated here... forever!

Could the three non-abstract attributes simply derive from TestCaseSourceAttribute? Or from an abstract base that we split out of TestCaseSourceAttribute?

I realize that this could seem like starting over, but it isn't really because the programmer is always made smarter by the code he writes, whether it gets used as is or not. IOW I think it could go pretty fast if we did a reset.

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub

acco32 commented:
@CharliePoole @rprouse 

I've made some modifications regarding extending the TestCaseSourceAttribute. Please have a look and let me know if this meets the standard. I am still adding more tests as I go along but thought this might be a good checkpoint. Thanks.

CharliePoole commented:
Heading out on vacation today! @rprouse will take care of your PR. :-)

acco32 commented:
@rprouse 
I've done quite a bit a re-factoring with additional tests. Please review when ready. Thanks.

@CharliePoole 
Enjoy your vacation.

rprouse commented:
The build failure on Travis looks like an unrelated Mono 3.2.8 crash. I have restarted that build.

rprouse commented on nunit.sln:
```diff
@@ -258,6 +272,10 @@ Global
 		{08D0F53A-CD37-4060-B5A2-BCC202F0DE46}.Debug|Any CPU.Build.0 = Debug|Any CPU
 		{08D0F53A-CD37-4060-B5A2-BCC202F0DE46}.Release|Any CPU.ActiveCfg = Release|Any CPU
 		{08D0F53A-CD37-4060-B5A2-BCC202F0DE46}.Release|Any CPU.Build.0 = Release|Any CPU
+		{DBD0C8B0-BE4E-4CBF-AAD4-44BEA2EA830B}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
```
All of the changes in the solution just look like MonoDevelop re-arranging things. You might want to revert all of your changes to this file otherwise merging in latest from master is going to be impossible.

rprouse commented on src/NUnitFramework/framework/Attributes/CsvDataSourceAttribute.cs:
```diff
@@ -0,0 +1,104 @@
+    /// <summary>
+    /// Imports CSV data from a file
+    /// </summary>
+    public class CsvDataAttribute : TestCaseSourceAttribute
```
I like this approach much better than your original approach. :+1:

rprouse commented on src/NUnitFramework/framework/Attributes/CsvDataSourceAttribute.cs:
```diff
@@ -0,0 +1,104 @@
+    /// </summary>
+    public class CsvDataAttribute : TestCaseSourceAttribute
+    {
+        private CsvData csv;
```
Our coding standard is that member variables start with an underscore, `_csv`

acco32 replied:
Okay. Do you guys publish the coding standard anywhere?

rprouse replied:
Coding standard is with the docs on the wiki at https://github.com/nunit/docs/wiki/Coding-Standards

rprouse commented on src/NUnitFramework/framework/Attributes/CsvDataSourceAttribute.cs:
```diff
@@ -0,0 +1,104 @@
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        /// <param name="source"></param>
```
I assume that source is the filename, but we are going to need much more complete XML Docs for public facing classes. That goes for all of your doc comments.

rprouse commented on src/NUnitFramework/framework/Attributes/TestCaseSourceAttribute.cs:
```diff
@@ -243,7 +243,7 @@ private IEnumerable<ITestCaseData> GetTestCasesFor(IMethodInfo method)
         }
 
-        private IEnumerable GetTestCaseSource(IMethodInfo method)
+        protected virtual IEnumerable GetTestCaseSource(IMethodInfo method)
```
One thing that bothers me with the approach of overriding the `TestCaseSourceAttribute` is that you are not calling to this base from the derived classes. Because of that, I don't feel that this is quite the right implementation. I think that if you want to take this approach, then you should move everything in `TestCaseSourceAttribute` into a new abstract base class `BaseTestCaseSourceAttribute` making the `GetTestCaseSource` method abstract. Then each of the source attributes would derive from that.

rprouse commented on src/NUnitFramework/framework/Attributes/DataSourceAttribute.cs:
```diff
@@ -0,0 +1,35 @@
+    /// <summary>
+    /// 
+    /// </summary>
+    public class DataSourceAttribute
```
Is this class still used? Shouldn't it be deleted?

acco32 replied:
Good catch. Not used anymore as I split it into separate classes per source.

rprouse commented on src/NUnitFramework/framework/Attributes/DatabaseDataSourceAttribute.cs:
```diff
@@ -0,0 +1,102 @@
+            get { return db.RowsToRead; }
+            set { db.RowsToRead = value; }
+        }
+
```
An OCD pet-peeve of mine is extra blank lines where they are not needed. Please clean up :smile:

rprouse commented on src/NUnitFramework/framework/Attributes/XmlDataSourceAttribute.cs:
```diff
@@ -0,0 +1,104 @@
+            var dataFromQuery = xml.GetData();
+
+            // check if items were returned
+            if (((IList)dataFromQuery).Count == 0)
```
I don't like that you have internal knowledge that the returned data is an IList. Why not just switch the return type to be IList?

rprouse commented on src/NUnitFramework/framework/Attributes/XmlDataSourceAttribute.cs:
```diff
@@ -0,0 +1,104 @@
+            if (!xml.SourceFileExists)
+            {
+                string errMsg = String.Format("Unable to find source: {0}", xml.Name);
+                throw new FileNotFoundException(errMsg);
```
You shouldn't be throwing these exceptions here or below (in each of your classes), but returning non-runnable tests. See how it is done in `TestCaseSourceAttribute`. For example, the method `ReturnErrorAsParameter`.

rprouse commented on src/NUnitFramework/framework/CsvData.cs:
```diff
@@ -0,0 +1,113 @@
+﻿// ***********************************************************************
+// Copyright (c) 2009-2015 Charlie Poole
```
Don't copy the year. New files should just be 2016.

rprouse commented on src/NUnitFramework/framework/CsvData.cs:
```diff
@@ -0,0 +1,113 @@
+                        continue;
+
+                    // once we have a valid (non-empty) string I continue to process
+                    var csv = lineData.Split(',');
```
This is going to split on commas within the data, for example, 

```
1,"To be, or not to be"
2,"No commas"
```

First row will be split in 3, second in two.

acco32 replied:
Added a new parameter for a user defined delimiter. This will help in part. Will need to put more logic regarding encountering items in quotes.

rprouse commented on src/NUnitFramework/framework/CsvData.cs:
```diff
@@ -0,0 +1,113 @@
+                        break;
+
+                    // skip empty lines as denoted by a newline character
+                    if (lineData == string.Empty || lineData == Environment.NewLine)
```
You should probably use `string.IsNullOrWhitespace` here and you don't need to check for the NewLine because the string that is returned from `ReadLine` does not contain the terminating carriage return or line feed.

oznetmaster replied:
`String.IsNullOrWhitespace` is a .NET 4.0 feature.  It should not be used if we want to maintain .NET 2.0 compatibility.

rprouse commented on src/NUnitFramework/framework/CsvData.cs:
```diff
@@ -0,0 +1,113 @@
+        /// </summary>
+        /// <param name="s">identified csv row</param>
+        /// <returns>array with extra characters removed</returns>
+        private string[] stripExtraCharacters(string[] s)
```
This trims extra characters at the start and end, but I don't think it unescapes characters within the string, primarily a `\"`

rprouse commented on src/NUnitFramework/framework/DataSource.cs:
```diff
@@ -0,0 +1,62 @@
+            set
+            {
+                if (value < 0)
+                    throw new Exception("Cannot set Rows to Read less than zero");
```
This should probably be an `ArgumentException`

rprouse commented on src/NUnitFramework/framework/DatabaseData.cs:
```diff
@@ -0,0 +1,382 @@
+            if (drivers == null)
+                return new Tuple<bool, string>(false, string.Empty);
+
+            foreach (var dvr in drivers)
```
Why are you verifying this way? Why not just fail when they fail to connect using the given connection string. You are limiting this to working with a hard coded list of drivers and you are missing quite a few common ones like SQLite and I think SQL Server LocalDB.

acco32 replied:
I only wanted to do the drivers that are bundled with the framework as a first iteration.

rprouse commented on src/NUnitFramework/framework/nunit.framework-4.5.csproj:
```diff
@@ -291,6 +294,9 @@
     <Compile Include="Constraints\UniqueItemsConstraint.cs" />
     <Compile Include="Constraints\XmlSerializableConstraint.cs" />
     <Compile Include="Contains.cs" />
+    <Compile Include="CsvData.cs" />
```
Obviously, you will need to add these to all of the other targets. Probably easier to do once you merge latest from master.

rprouse commented:
Some final comments after reviewing,

1. You will likely want to merge in latest from master. I think you are well behind master.
2. You need a wider variety of CVS test cases. I would like to see files with complex strings with things like commas and quotes within the strings.
3. The database logic seems more complicated than it needs to be and limits you to known drivers
4. Once you get everything up to snuff, you should be adding equivalent `TestFixtureSource` attributes.
5. Once you expand to other platforms, you will need to determine which of these works on each platform.
6. Don't forget the CF projects.

acco32 commented:
@rprouse 
Thanks for the review. I'll update and submit again. A few questions:

- In order to add to other projects, would I move these files to an earlier version of product as those files are shared with later versions? And how far back (perhaps 3.0)?
- What areas of the database logic are confusing? Perhaps we can go over this more in detail.
- I've not worked with the CF projects. I've noticed in some places some preprocessor directives for that project. Any guidelines I should following before messing around with those projects?
- Perhaps *TestFixtureSource* equivalents can be a separate pull request. I'd like to get this one to level of completion to merge (been working on this for quite some time).

Thanks.

rprouse commented:
>In order to add to other projects, would I move these files to an earlier version of product as those files are shared with later versions? And how far back (perhaps 3.0)?

You will find that each of the projects are in the same directory, so to add your files to the other target frameworks, just expand the project, click the show all files button, then right click on your files and include in project.

rprouse commented:
>I've not worked with the CF projects. I've noticed in some places some preprocessor directives for that project. Any guidelines I should following before messing around with those projects?

I am not sure if we would want to include these features in CF. @oznetmaster what do you think?

oznetmaster commented:
CF applications can use both external sql databases, sqlite and sql server compact.

It seems like this enhancement should apply to the CF build as well.

acco32 commented:
@rprouse 
I'm trying to track down the latest error with the build on AppVeyor. It looks like it may be a Cake error but I am not sure.

CharliePoole commented:
You have a test failure in the .NET 4.0 framework tests. See https://ci.appveyor.com/project/CharliePoole/nunit/build/3.5.0-CI-3068-PR-843#L757

CharliePoole commented:
It's desirable, but not absolutely necessary, to merge this before splitting the repositories as described in #1598. The PR deals only with the framework.

rprouse commented:
@acco32 I am going to close this PR. It has been ongoing for over a year now. 

I also think that this feature might be better as an extension to NUnit. If you want to continue to work on it, I would suggest setting up a repo for this feature as an extension. If you need help, give me a shout.
